src/rust/ide/view/graph-editor/src/selection.rs

```diff
@@ -0,0 +1,510 @@
+    }
+
+    fn contains_y(&self, y:f32) -> bool {
+        y > self.bottom && y < self.top
+    }
+
+    pub fn width(&self) -> f32 {
+        self.right - self.left
+    }
+
+    pub fn height(&self) -> f32 {
+        self.top - self.bottom
+    }
+
+    pub fn intersects(&self, other:&BoundingBox) -> bool {
+        // https://stackoverflow.com/a/13390495
+        // self.right<other.left or other.right<self.left or self.bottom<other.top or other.bottom<self.top)
+        let not_contained = (self.right < other.left)
+                         || (other.right < self.left)
+                         || (self.bottom > other.top)
+                         || (other.bottom>self.top);
+        !not_contained
+    }
+}
+
+
+// === Bounding Box Tests ===
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_intersection() {
+        let bb1 = BoundingBox::from_corners(Vector2::new(0.5,0.5),Vector2::new(1.0,1.0));
+        let bb2 = BoundingBox::from_corners(Vector2::new(0.0,0.0),Vector2::new(2.0,2.0));
+        assert!(bb1.intersects(&bb2));
+        assert!(bb2.intersects(&bb1));
+
+        let bb1 = BoundingBox::from_corners(Vector2::new(3.0,3.0),Vector2::new(4.0,4.0));
+        let bb2 = BoundingBox::from_corners(Vector2::new(0.0,0.0),Vector2::new(2.0,2.0));
+        assert!(!bb1.intersects(&bb2));
+        assert!(!bb2.intersects(&bb1));
+
+        let bb1 = BoundingBox::from_corners(Vector2::new(0.0,0.0),Vector2::new(4.0,4.0));
+        let bb2 = BoundingBox::from_corners(Vector2::new(0.0,0.0),Vector2::new(-2.0,-2.0));
+        assert!(bb1.intersects(&bb2));
+        assert!(bb2.intersects(&bb1));
+
+        let bb1 = BoundingBox::from_corners(Vector2::new(0.0,0.0),Vector2::new(4.0,4.0));
+        let bb2 = BoundingBox::from_corners(Vector2::new(2.0,2.0),Vector2::new(200.0,200.0));
+        assert!(bb1.intersects(&bb2));
+        assert!(bb2.intersects(&bb1));
+
+        let bb1 = BoundingBox::from_corners(Vector2::new(-50.0,-50.0),Vector2::new(25.0,25.0));
+        let bb2 = BoundingBox::from_corners(Vector2::new(5.00,50.0),Vector2::new(100.0,100.0));
+        assert!(!bb1.intersects(&bb2));
+        assert!(!bb2.intersects(&bb1));
+    }
+}
+
+
+// ==========================
+// === TemporarySelection ===
+// ==========================
+
+/// Struct that stores the initial selection state of a node. Used to restore the selection state
+/// after a temporary selection has been undone, for example when an area selection is shrunk to no
+/// longer encompass the node.
+///
+/// This implements `Hash` and `Eq` based only on the ndoe identity to avoid overwriting the initial
+/// selection state in a set of `TemporarySelection`s. This allows the first added element for a
+/// node to be preserved.
+#[derive(Clone,Copy,Default,Debug,Eq)]
+pub struct TemporarySelection {
+    node         : NodeId,
+    was_selected : bool,
+}
+
+impl TemporarySelection {
+    fn new(node:NodeId,was_selected:bool)  -> Self {
```

Comment:
```suggestion
    fn new(node:NodeId, was_selected:bool)  -> Self {
```
And perhaps we could make it public?